### ISSUE TITLE
Memory error on Vega 10

### ISSUE THREAD

**Issue by biswagsingh** [plaidml/plaidml#34]
Hi I am trying plaid ml on AMD Vega 10 : gfx900

I get the following error:

prj47-rack-06@PRJ47-RACK-06:~/biswa/plaidbench$ python plaidbench.py mobilenet
Using PlaidML backend.
INFO:plaidml:Initializing device gfx900.0: "gfx900", vendor "Advanced Micro Devi                                                                             ces, Inc."
INFO:plaidml:Initializing device gfx900.1: "gfx900", vendor "Advanced Micro Devi                                                                             ces, Inc."
INFO:plaidml:Initializing device gfx900.2: "gfx900", vendor "Advanced Micro Devi                                                                             ces, Inc."
INFO:plaidml:Initializing device gfx900.3: "gfx900", vendor "Advanced Micro Devi                                                                             ces, Inc."
INFO:plaidml:Opening device "gfx900.3": "Advanced Micro Devices, Inc. gfx900"

Model loaded.
Compiling and running initial batch, batch_size=1
Warmup
Memory access fault by GPU node-7 on address 0x4408bd6000. Reason: Page not pres                                                                             ent or supervisor privilege.
Aborted (core dumped)

Any idea how to resolve this?

Thanks,
Biswa

**Comment by jamilbk** [comment]
@biswagsingh I hit the same error. It also happens for me when doing MultiGPU training with [hipCaffe](https://github.com/ROCmSoftwarePlatform/hipCaffe) on ROCm 1.6.4 so I'm not entirely sure it's an issue with PlaidML. Specifying a batch size of 100 prevents the segfault from happening on my box, but then causes a hard kernel freeze so it's not exactly a workaround.

This segfault only happens when I've selected a different OpenCL device than the first (i.e. the one with the `0` index). This is the cause of the hipCaffe segfault as well.

Try editing your ~/.plaidml and ensuring the first OpenCL device(`gfx900.0`) is selected:

```json
{
    "PLAIDML_DEVICE_IDS":[
      "gfx900.0"
    ],
    "PLAIDML_EXPERIMENTAL":true,
    "PLAIDML_TELEMETRY":true
}
```

Hope that helps.

**Comment by brianretford** [comment]
@biswagsingh what is your exact OS version & distro? What CPU model do you have?
Can you try running clpeak and see if your problem repros?
Either:
`bazel run //external:clpeak`
or
```
sudo apt install clpeak
clpeak
```

Our Vega 10 works but I think we have driver version 1.6.3. One thing we've noticed about the Vega 10 is that they are *very* particular about the amount of power they have. They also require a fairly modern CPU (specifically ones that support PCIe atomics -- see the ROCm site).

**Comment by brianretford** [comment]
Same questions to you @jamilbk

**Comment by jamilbk** [comment]
| CPU | Threadripper 1950X |
| RAM | 128 GB G.Skill DDR4 @ 2933 CL14 |
| Board | Gigabyte X399 Gaming 7 |
| GPU | 4x water cooled Vega FE |
| PSU | EVGA 1600 W SuperNova P2 |

I'm fairly certain my PSU is not the issue since I've stress-tested this machine for many hours in LuxMark, Furmark, and Prime95 measuring up to 1600W draw at the wall with no stability issues.

FWIW, [here is a potentially relevant bug in MIOpen](https://github.com/ROCmSoftwarePlatform/hipCaffe/issues/11) that I hit as well trying to train with hipCaffe and MultiGPU. AMD's suggestion did not work for me and I ended up giving up.

[clinfo.txt](https://github.com/plaidml/plaidml/files/1409561/clinfo.txt)
[clpeak.txt](https://github.com/plaidml/plaidml/files/1409562/clpeak.txt)
[uname.txt](https://github.com/plaidml/plaidml/files/1409563/uname.txt)

**Comment by brianretford** [comment]
Very useful, thanks. Great to note that double rate half precision requires
explicit vectorization.

**Comment by biswagsingh** [comment]
Thank you guys for the reply
@jamilbk while setting up plaidml I gave {
    "PLAIDML_DEVICE_IDS":[
        "gfx900.3"
    ],
    "PLAIDML_EXPERIMENTAL":true,
    "PLAIDML_TELEMETRY":false
}

@brianretford 
Our Distro is below

prj47-rack-06@PRJ47-RACK-06:~/biswa/plaidbench$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.3 LTS
Release:        16.04
Codename:       xenial

Our CPU is  AMD EPYC 7601 32-Core Processor

I got it working for miopen hiptensorflow and wanted to compare with plaidml.

I am not able to run clpeak with following problem

ERROR: Skipping '//external:clpeak': no such target '//external:clpeak': target 'clpeak' not declared in package 'external' defined by /home/prj47-rack-06/biswa/plaidbench/WORKSPACE
WARNING: Target pattern parsing failed.
ERROR: no such target '//external:clpeak': target 'clpeak' not declared in package 'external' defined by /home/prj47-rack-06/biswa/plaidbench/WORKSPACE

Can you help me resolve this?

Thanks,
Biswa

**Comment by jamilbk** [comment]
@biswagsingh Try building clpeak from source:

```
git clone https://github.com/krrishnarraj/clpeak.git
cd clpeak
mkdir build && cd build
cmake ../
make
./clpeak
```

Also, if you change your chosen graphics device to `gfx900.0` instead of `gfx900.3` in ~/.plaidml I bet that will fix the issue for now at least.

**Comment by brianretford** [comment]
You have to build clpeak from plaidml, not plaidbench. I'll make a note to
add it to both of them though.

**Comment by biswagsingh** [comment]
@jamilbk  Thanks for clpeak instruction. Also I changed to gfx900.0 , but unfortunately it doesn't help. Same error with mobilenet 

The clpeak log is attached @brianretford @jamilbk 
[clpeak.txt](https://github.com/plaidml/plaidml/files/1409620/clpeak.txt)
Thanks,
Biswa

**Comment by biswagsingh** [comment]
@jamilbk @brianretford 

Following is the hiptensorflow result for your reference on the same machine

prj47-rack-06@PRJ47-RACK-06:~/biswa/plaidbench$ python plaidbench.py mobilenet --no-plaid
Using TensorFlow backend.
I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libhipblas.so locally
I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libhipfft.so locally
I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libhip_hcc.so locally
I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libhiprng.so locally
I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libMIOpen.so locally
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/stream_executor/cuda/cuda_driver.cc:633] creating context when one is currently active; existing: 0x262ad40
I tensorflow/core/common_runtime/gpu/gpu_device.cc:892] Found device 0 with properties:
name: Device 6860
major: 3 minor: 0 memoryClockRate (GHz) 1.5
pciBusID 0000:06:00.0
Total memory: 15.98GiB
Free memory: 15.73GiB
I tensorflow/core/common_runtime/gpu/gpu_device.cc:913] DMA: 0
I tensorflow/core/common_runtime/gpu/gpu_device.cc:923] 0:   Y
I tensorflow/core/common_runtime/gpu/gpu_device.cc:972] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Device 6860, pci bus id: 0000:06:00.0)

Model loaded.
Compiling and running initial batch, batch_size=1
I tensorflow/core/kernels/conv_ops.cc:604] running auto-tune for Convolve
Warmup
Doing the main timing
Example finished, elapsed: 0.639880180359 (compile), 24.248714447 (execution)

Thanks,
Biswa

**Comment by biswagsingh** [comment]
@jamilbk @brianretford 
with --verbose it passes at times and fails most of the time. Bot he logs are attached
[log_error.txt](https://github.com/plaidml/plaidml/files/1409728/log_error.txt)

**Comment by brianretford** [comment]
Great, thanks. I just updated all of our drivers and I'm still not able to get a repro, though we do get 22TFLOPs of fp16. Plaidbench output for us:
```
vertexai@lab-06:~/plaidbench$ sudo python plaidbench.py mobilenet [installe
Using PlaidML backend.
INFO:plaidml:Initializing device gfx900.0: "gfx900", vendor "Advanced Micro Devices, Inc."
INFO:plaidml:Opening device "gfx900.0": "Advanced Micro Devices, Inc. gfx900"

Model loaded.
Compiling and running initial batch, batch_size=1
Warmup
Doing the main timing
Example finished, elapsed: 6.91148900986 (compile), 5.12591791153 (execution)
```

What's the output of `apt list --installed | grep rocm`? Ours:
```
Linux lab-06 4.11.0-kfd-compute-rocm-rel-1.6-148 #1 SMP Wed Aug 23 12:00:35 CDT 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by brianretford** [comment]
Really good to know about verbose though. We have a fix for one race condition we'll be pushing tomorrow.

**Comment by biswagsingh** [comment]
@brianretford 
prj47-rack-06@PRJ47-RACK-06:~$ uname -a
Linux PRJ47-RACK-06 4.11.0-kfd-compute-rocm-rel-1.6-180 #1 SMP Tue Oct 10 08:15:38 CDT 2017 x86_64 x86_64 x86_64 GNU/Linux

prj47-rack-06@PRJ47-RACK-06:~$ apt list --installed | grep rocm/Ubuntu

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

rocm/Ubuntu 16.04,now 1.6.180 amd64 [installed]


Yes --verbose passes the test at times. Will wait for the race condition to be resolved and will try the test again. 
Please let me know about the progress.

Thanks,
Biswa

**Comment by jamilbk** [comment]
@brianretford What are your system details? I think you're still stuck on ROCm 1.6.3 (`4.11.0-kfd-compute-rocm-rel-1.6-148`). That should be `4.11.0-kfd-compute-rocm-rel-1.6-180` if you're on ROCm 1.6.4.

Would it be safe to assume the bug only surfaces when more than 1 Vega is present? (p.s. RX Vega 56 are [now $420 on the NewEgg Ebay Store](https://www.ebay.com/itm/SAPPHIRE-Radeon-RX-Vega-56-DirectX-12-100420L-8GB-2048-Bit-HBM2-CrossFireX-Suppo-/382242047727?hash=item58ff66c6ef))

For me the bug occurs with `gfx900.1`, `gfx900.2`, `gfx900.3`, but not `gfx900.0`. I tried changing the `Initial Display Output` setting in my BIOS to PCIe slot 2 and connected my monitor to that card, but the failure pattern with the above device IDs persists. I'm not sure how CL enumerates device IDs? Are these guaranteed to be always assigned to the same physical card, like hard drive UUIDs? If so, and `gfx900.0` always points to the same physical card, that may help narrow down the issue. If not, and `gfx900.0` points to whichever card is first initialized (i.e. from the BIOS setting) or has a display attached, then it could be related to that.

If it helps at all, I've attached the output of `dmesg | grep drm` below.

[dmesg.txt](https://github.com/plaidml/plaidml/files/1411412/dmesg.txt)

**Comment by brianretford** [comment]
I just rebooted and am now at 1.6.180 and things work fine. We definitely
want to build a 4x vega rig but haven't gotten around to it (will soon). I
have no idea how the opencl drivers enumerate their devices, as it's up to
each vendor to do that. @biswagsignh, could you try each of your cards and
see if one of them works? I doubt it at this point. I think it's probably a
reflection of a race condition or other issue in our code that manifests
inconsistently.

**Comment by biswagsingh** [comment]
@brianretford 
I see the issue with all the gpus. Hence I am also suspecting race condition issue as with --verbose it passes at times.

Thanks,
Biswa

**Comment by jamilbk** [comment]
Just as another data point, the `--verbose` flag does *not* affect my runs. Still fails consistently each time unless `gfx900.0` is chosen.

```
DEBUG:plaidml:Launching tile program: kernel_c13_sdk_1 global(16384, 1, 1) local(256, 1, 1) outputs(X_T8)inputs(X_T6)
DEBUG:plaidml:  Output: 1048576 bytes in HAL Buffer 0x239c558
DEBUG:plaidml:  TmpInput tidx=0
DEBUG:plaidml:Ran kernel_c13_sdk_0: dur=4.9927e-05 GFL/s=5.25055e+09 GBP/s= 2.10022e+10
DEBUG:plaidml:Ran kernel_c13_sdk_1: dur=1.4371e-05 GFL/s=3.64824e+10 GBP/s= 7.52449e+10
Memory access fault by GPU node-3 on address 0x2402a13000. Reason: Page not present or supervisor privilege.
[1]    53542 abort (core dumped)  python plaidbench.py mobilenet --verbose
```

**Comment by jamilbk** [comment]
Interestingly, I'm able to run all the benchmarks just fine on any GPU with `--train`.

**Comment by earhart** [comment]
@jamilbk That's a pretty interesting error -- thanks for letting us know!

Is there any chance you could run with `--verbose=4`, and mail us the resulting output?  It'll print out a _ton_ of low-level information; it'll be a lot to go through, but it would really help us narrow down what's going wrong.

**Comment by jamilbk** [comment]
Sure thing. I've attached the output here in case others want to follow along.

[crash.txt.gz](https://github.com/plaidml/plaidml/files/1415399/crash.txt.gz)

**Comment by earhart** [comment]
Just a quick update -- looking at the ROC drivers, the only way I see to get this sort of error is by accessing system physical addresses from the GPU, when going through an IOMMU, where the addresses haven't been mapped into the GPU's address space.  I've looked through the crash log and verified the kernels by hand; they look fine to me, so I don't think the code is walking off the end of a buffer.

I'm looking at the way we set up and synchronize memory transfers; that seems like the most likely culprit.  We're also trying to reproduce the problem on one of our machines, which'll make debugging it a lot easier.

Thanks for the crash log!

**Comment by brianretford** [comment]
We released 0.1.1 -- There's some small chance that it will resolve these issues, though I doubt it. It's worth giving it a try and seeing if it does. We're still investigating over here.

**Comment by jamilbk** [comment]
The issue persists for me on `0.1.1`. Further, I'm now getting this error on `gfx900.0` with `--train --fp16` whereas before I was [getting an OpenCL error](https://github.com/plaidml/plaidbench/issues/17)

**Comment by jamilbk** [comment]
Here's the debug log from `python plaidbench.py --train --fp16 mobilenet --verbose=4` on `gfx900.0` in case it's useful.

[crash-train.txt.gz](https://github.com/plaidml/plaidml/files/1428508/crash-train.txt.gz)

**Comment by brianretford** [comment]
If you're still interested in testing this, we've changed a lot of things in how memory works in 0.1.4. Would be interested to see if this still reproduces

**Comment by tzerrell** [comment]
We have made various memory management changes since this was reported; since we haven't had a reproduction reported, we're closing this. Please re-open or make a new issue if you reproduce this!
